server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -226,61 +279,149 @@ public float getShardBalance() {
+            for (IndexMetadata indexMetadata : metadata.indices().values()) {
+                writeLoad += getIndexWriteLoad(writeLoadForecaster, indexMetadata);
+            }
+            return writeLoad;
+        }
+
+        private static double getIndexWriteLoad(WriteLoadForecaster writeLoadForecaster, IndexMetadata indexMetadata) {
+            var shardWriteLoad = writeLoadForecaster.getForecastedWriteLoad(indexMetadata).orElse(0.0);
+            return shardWriteLoad * numberOfCopies(indexMetadata);
+        }
+
+        private static long getTotalDiskUsageInBytes(ClusterInfoService clusterInfoService, Metadata metadata) {
+            long totalDiskUsageInBytes = 0;
+            for (IndexMetadata indexMetadata : metadata.indices().values()) {
+                totalDiskUsageInBytes += getIndexDiskUsageInBytes(clusterInfoService, indexMetadata);
+            }
+            return totalDiskUsageInBytes;
+        }
+
+        private static long getIndexDiskUsageInBytes(ClusterInfoService clusterInfoService, IndexMetadata indexMetadata) {
+            OptionalLong forecastedShardSizeInBytes = indexMetadata.getForecastedShardSizeInBytes();
+            final long indexDiskUsageInBytes;
+            if (forecastedShardSizeInBytes.isPresent()) {
+                indexDiskUsageInBytes = forecastedShardSizeInBytes.getAsLong() * numberOfCopies(indexMetadata);
+            } else {
+                indexDiskUsageInBytes = getIndexDiskUsageInBytesFromClusterInfo(clusterInfoService, indexMetadata);
+            }
+            return indexDiskUsageInBytes;
+        }
+
+        private static long getIndexDiskUsageInBytesFromClusterInfo(ClusterInfoService clusterInfoService, IndexMetadata indexMetadata) {
+            long totalSizeInBytes = 0;
+            int shardCount = 0;
+            final ClusterInfo clusterInfo = clusterInfoService.getClusterInfo();
+            for (int shard = 0; shard < indexMetadata.getNumberOfShards(); shard++) {
+                final ShardId shardId = new ShardId(indexMetadata.getIndex(), shard);
+
+                final Long primaryShardSize = clusterInfo.getShardSize(shardId, true);
+                if (primaryShardSize != null) {
+                    totalSizeInBytes += primaryShardSize;
+                    shardCount++;
+                }
+
+                final Long replicaShardSize = clusterInfo.getShardSize(shardId, false);
+                if (replicaShardSize != null) {
+                    totalSizeInBytes += (replicaShardSize * indexMetadata.getNumberOfReplicas());
+                    shardCount += indexMetadata.getNumberOfReplicas();
+                }
+            }
+
+            if (shardCount == numberOfCopies(indexMetadata)) {
+                return totalSizeInBytes;
+            }
+
+            // TODO: Should we go through the cluster info service and compute the average in this case?
+            return shardCount == 0 ? 0 : (totalSizeInBytes / shardCount) * numberOfCopies(indexMetadata);
```

Comment:
@henningandersen do you think that computing the average with the available shards and then multiplying by the number of copies is the right call here? I think it's fine in most cases, but I just want to confirm that my intuition is correct.

Reply:
Yeah, that sounds fine to me.

I wonder if we should also ask the cluster info though in case there is no forecasted shard size in `diskUsageInBytesPerShard`? DIfferent topic, also fine to do in a follow-up ofc.